### DISCUSSION TITLE
Unsure how to use `unsafeType` within `defineNodeType`

### DISCUSSION THREAD

**Discussion by Mearman** [breadboard-ai/breadboard#1194] [Q&A] [answered]
When trying to use the examples from the readme, it still only allows the primitives, so I get: "`Type '"object"' is not assignable to type '"string" | "number" | "boolean"`".
Is there a step I'm missing? It would be useful to see a sample of how to use `unsafeType` or in general a non-primitive with `defineNodeType`.

https://github.com/breadboard-ai/breadboard/blob/a20538fe046a5a6258ff7b2bafd8244d3015d275/packages/build/README.md?plain=1#L101-L109

this is a (currently invalid) sample of what I was attempting. without the `unsafeType` for clarity
```typescript
export const entries = defineNodeType({
  inputs: {
    object: {
      type: "object",
      description: "The object to get entries from",
    },
  },
  outputs: {
    entries: {
      type: "array",
      description: "The entries of the object",
      primary: true,
    },
  },
  invoke: ({ object }) => {
    return {
      entries: Object.entries(object),
    };
  },
});
```

**Comment by Mearman** [comment]
- see https://github.com/breadboard-ai/breadboard/pull/1197

**Comment by aomarks** [comment]
> When trying to use the examples from the readme, it still only allows the primitives, so I get: "`Type '"object"' is not assignable to type '"string" | "number" | "boolean"`".

I think you just need to sync with `main` to pull in https://github.com/breadboard-ai/breadboard/pull/1196. Among other things, that PR expanded `unsafeType` to allow taking any valid JSON Schema, whereas previously it was a small subset.

**Reply by Mearman** [reply]
will have a crack at it again with the new changes. though I have a feeling [this](https://github.com/orgs/breadboard-ai/discussions/1195) is going to block me next

**Reply by Mearman** [reply]
@aomarks I might need pointing in the right direction in terms of how to handle an `object` or `array` of unknowns.
here's an example.
how would the `inputs` and `outputs` be written for `defineNodeType` with this function:
```typescript
async function invokePop<T>({ array }: { array: T[] }): Promise<{
  item: T | undefined;
  array: T[];
}> {
  const item = array.pop();
  return {
    item,
    array,
  };
}
```

**Reply by aomarks** [reply]
I just added `"unknown"` in https://github.com/breadboard-ai/breadboard/pull/1200, so you should now be able to do something like:

```ts
import { defineNodeType, array } from "@breadboard-ai/build";

export const pop = defineNodeType({
  inputs: {
    arr: {
      type: array("unknown"),
    },
  },
  outputs: {
    last: {
      type: "unknown"
    },
  },
  invoke: ({ arr }) => ({
    last: arr.pop(),
  }),
});
```

This isn't going to carry the generic information, but that should be fine for now and and I'll have a think about how we can express it in a way that does carry the generic.

**Reply by Mearman** [reply]
superb. I shall experiment.
I was also pondering about the concept of generics in BGL. I think it'd certainly be useful in cases like the [boards that require a predicate](https://github.com/orgs/breadboard-ai/discussions/1195#discussioncomment-8896989).

**Reply by Mearman** [reply, marked as answer]
we edge closer.
~to export, I need to type the const. though I can't figure out the correct config for the `MonomorphicDefinition` generics.~
scratch that. I think I've got it
```typescript
export const popNodeType: MonomorphicDefinition<
  {
    array: {
      type: AdvancedBreadboardType<unknown[]>;
    };
  },
  {
    last: {
      type: "unknown";
    };
    array: {
      type: AdvancedBreadboardType<unknown[]>;
    };
  }
> = defineNodeType({
  inputs: {
    array: {
      type: array("unknown"),
    },
  },
  outputs: {
    last: {
      type: "unknown",
    },
    array: {
      type: array("unknown"),
    },
  },
  invoke: <T>(inputs: { array: T[] }): { last: T | undefined; array: T[] } => {
    const { array } = inputs;
    const last = array.pop();
    return {
      last,
      array,
    };
  },
});
```

**Comment by Mearman** [comment]
Leaving these here for reference
```typescript
export const popNodeType = defineNodeType({
  inputs: {
    array: {
      type: array("unknown"),
    },
  },
  outputs: {
    last: {
      type: "unknown",
    },
    array: {
      type: array("unknown"),
    },
  },
  invoke: <T>(inputs: { array: T[] }): { last: T | undefined; array: T[] } => {
    const { array } = inputs;
    const last = array.pop();
    return {
      last,
      array,
    };
  },
});
```
```typescript
const popNodeHandler: NodeHandler = {
  invoke: <T>({
    array,
  }: {
    array: T[];
  }): Promise<{ last: T | undefined; array: T[] }> => {
    const last = array.pop();
    return Promise.resolve({
      last,
      array,
    });
  },
};
```
